test/net/sourceforge/kolmafia/session/LimitModeTest.java

```diff
@@ -394,4 +433,175 @@ void someLimitMCD(final LimitMode lm) {
       assertThat(lm.limitMCD(), is(true));
     }
   }
+
+  @Nested
+  class Recovery {
+    @ParameterizedTest
+    @EnumSource(
+        value = LimitMode.class,
+        names = {"NONE", "ASTRAL", "BIRD", "MOLE", "ROACH"})
+    void someDontLimitRecovery(final LimitMode lm) {
+      assertThat(lm.limitRecovery(), is(false));
+    }
+
+    @ParameterizedTest
+    @EnumSource(
+        value = LimitMode.class,
+        names = {"SPELUNKY", "BATMAN", "ED"})
+    void someLimitRecovery(final LimitMode lm) {
+      assertThat(lm.limitRecovery(), is(true));
+    }
+  }
+
+  // *** Pseudo LimitModes
+
+  @Nested
+  class Astral {
+    @AfterEach
+    public void afterEach() {
+      KoLConstants.activeEffects.clear();
+    }
+
+    private static final AdventureResult ASTRAL_MUSHROOM =
+        ItemPool.get(ItemPool.ASTRAL_MUSHROOM, 1);
+    private static final AdventureResult HALF_ASTRAL = EffectPool.get(EffectPool.HALF_ASTRAL);
+
+    @Test
+    void gainingHalfAstralEntersAstralLimitMode() {
+      var builder = new FakeHttpClientBuilder();
+      var client = builder.client;
+      var cleanups =
+          new Cleanups(
+              withHttpClientBuilder(builder),
+              withItem(ASTRAL_MUSHROOM),
+              withNoEffects(),
+              withProperty("currentAstralTrip", ""),
+              withLimitMode(LimitMode.NONE));
+      try (cleanups) {
+        client.addResponse(200, html("request/test_use_astral_mushroom.html"));
+        client.addResponse(200, ""); // api.php
+
+        var request = UseItemRequest.getInstance(ASTRAL_MUSHROOM);
+        request.run();
+
+        assertEquals(5, HALF_ASTRAL.getCount(KoLConstants.activeEffects));
+        assertEquals(KoLCharacter.getLimitMode(), LimitMode.ASTRAL);
+
+        var requests = client.getRequests();
+        assertThat(requests, hasSize(2));
+
+        assertPostRequest(
+            requests.get(0), "/inv_use.php", "whichitem=" + ItemPool.ASTRAL_MUSHROOM + "&ajax=1");
+        assertPostRequest(requests.get(1), "/api.php", "what=status&for=KoLmafia");
+      }
+    }
+
+    @Test
+    void losingHalfAstralLeavesAstralLimitMode() {
+      var cleanups =
+          new Cleanups(
+              withEffect(EffectPool.HALF_ASTRAL, 1),
+              withProperty("currentAstralTrip", "Great Trip"),
+              withLimitMode(LimitMode.ASTRAL));
+      try (cleanups) {
+        AdventureResult adv = new AdventureResult(AdventureResult.ADV, -1);
+        ResultProcessor.processResult(true, adv);
+
+        assertEquals(0, HALF_ASTRAL.getCount(KoLConstants.activeEffects));
+        assertEquals(KoLCharacter.getLimitMode(), LimitMode.NONE);
+        assertThat(Preferences.getString("currentAstralTrip"), is(""));
+      }
+    }
+  }
+
+  @Nested
+  class Mole {
+    @AfterEach
+    public void afterEach() {
+      KoLConstants.activeEffects.clear();
```

Comment:
this should be unnecessary due to the cleanups that add and clear effects?

Reply:
Yeah. That was before I invented Player.withNoEffects